cs/src/Management/TunnelManagementClient.cs

```diff
@@ -1,1566 +1,1570 @@
+                            if (allowNotFound && response.StatusCode == HttpStatusCode.NotFound &&
+                                problemDetails.Detail == null)
+                            {
+                                return default;
+                            }
+
+                            errorMessage = "Tunnel service error: " +
+                                problemDetails!.Title + " " + problemDetails.Detail;
+                            if (problemDetails.Errors != null)
+                            {
+                                foreach (var error in problemDetails.Errors)
+                                {
+                                    var messages = string.Join(" ", error.Value);
+                                    errorMessage += $"\n{error.Key}: {messages}";
+                                }
+                            }
+                        }
+                    }
+                    else if ((int)response.StatusCode >= 500)
+                    {
+                        // 5xx status responses may include VS SaaS error details.
+                        var errorDetails = await response.Content.ReadFromJsonAsync<ErrorDetails>(
+                            JsonOptions, cancellation);
+                        if (!string.IsNullOrEmpty(errorDetails?.Message))
+                        {
+                            errorMessage = "Tunnel service error: " + errorDetails!.Message;
+                            if (!string.IsNullOrEmpty(errorDetails.StackTrace))
+                            {
+                                errorMessage += "\n" + errorDetails.StackTrace;
+                            }
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // A default error message will be filled in below.
+                    innerException = ex;
+                }
+            }
+
+            errorMessage ??= "Tunnel service response status code: " + response.StatusCode;
+
+            if (response.Headers.TryGetValues(RequestIdHeaderName, out var requestId))
+            {
+                errorMessage += $"\nRequest ID: {requestId.First()}";
+            }
+
+            try
+            {
+                response.EnsureSuccessStatusCode();
+            }
+            catch (HttpRequestException hrex)
+            {
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.BadRequest:
+                        throw new ArgumentException(errorMessage, hrex);
+
+                    case HttpStatusCode.Unauthorized:
+                    case HttpStatusCode.Forbidden:
+                        // Enterprise Policies
+                        if (response.Headers.Contains("X-Enterprise-Policy-Failure"))
+                        {
+                            var message = response.Content != null ? await response.Content.ReadAsStringAsync() : string.Empty;

```

Comment:
Does the dev tunnels service already send out JSON error details?

Reply:
As the `ErrorDetails` class is changing in the same PR, I assume not yet. Then will this code break when it encounters a string response from the service where it expects JSON?  
If the content type of the response will change, it can be used to distinguish between a string or JSON. Or wrap `JsonSerializer.Deserialize<ErrorDetails>(message)` in `try-catch` and if it fails, assume it's a string, not json.

Reply:
Yes, the service already sends out json error details with the new response header key `X-Enterprise-Policy-Failure`.

Reply:
The suggestion on the service side was to send the response in ProblemDetails rather than a string value. All 3 policies send ProblemDetails json object instead of a string. And given this specific parsing will happen only when Enterprise policy response header is present, the expected content should be in json format. I don't see an exception to this, unless the response for a new enterprise policy in the future is send in a different format.

Reply:
I added an additional check to make sure there is no exception if the string is not in a json format.